### PR TITLE
Migrate frontend tooling to Webpack

### PR DESCRIPTION
**Pull request by miketheman**
Strap in, this is exciting!

The TL,DR is that Webpack 5.x (and associated plugins) replaces Gulp 4.x (and associated plugins).

For reviewers: I've tried to split the commits cleanly enough to follow the train of changes. Let me know if you have questions!

--- 

# Migrate Gulp to Webpack

## Purpose

Replace the aging [Gulp](https://gulpjs.com/) task runner with [Webpack](https://webpack.js.org/).
Gulp hasn't aged well in the JavaScript/CSS world, with many plugins being abandoned and the ecosystem being generally unstable.
Our setup also does not perform as quickly as we want, as it rebuilds the universe on changes, taking ~1-2 minutes for each change, and does not match current day frontend development practices.

Webpack can also help us with bundle splitting, which will help us with performance.

Alternatives considered, and why not chosen:

* [Parcel](https://parceljs.org/) - Parcel is a zero configuration bundler, which is great for small projects, but not for our large project. It does appear widely used with **templated HTML** projects.
* [Turbopack](https://turbo.build/pack) - Turbopack does not yet support Sass, and is not yet stable, so is not a great candidate for a production build.

We also already use Webpack for our JavaScript sources, so it makes sense to continue to use it for other static assets thought its plugin ecosystem.

## Current Implementation  

Current Gulp task flow (output of `gulp --tasks`):

```shellsession
...
 └─┬ default
   └─┬ <series>
     └─┬ dist
       └─┬ <series>
         ├── clean
         ├─┬ <parallel>
         │ ├─┬ dist:fontawesome
         │ │ └─┬ <parallel>
         │ │   ├── dist:fontawesome:css
         │ │   └── dist:fontawesome:fonts
         │ ├── dist:css-ltr
         │ ├── dist:css-rtl
         │ ├── dist:noscript
         │ ├── dist:js
         │ ├── dist:admin:fonts
         │ ├── dist:admin:css
         │ ├── dist:admin:js
         │ └── dist:vendor
         ├── dist:images
         ├─┬ <parallel>
         │ ├── dist:manifest
         │ └── dist:admin:manifest
         └─┬ <parallel>
           ├─┬ dist:compress
           │ └─┬ <parallel>
           │   ├── dist:compress:gz
           │   └─┬ dist:compress:br
           │     └─┬ <parallel>
           │       ├── dist:compress:br:generic
           │       └── dist:compress:br:text
           └─┬ dist:admin:compress
             └─┬ <parallel>
               ├── dist:admin:compress:gz
               └─┬ dist:admin:compress:br
                 └─┬ <parallel>
                   ├── dist:admin:compress:br:generic
                   └── dist:admin:compress:br:text
```

Breaking this down into natural language, here's the flow of steps (ignoring parallel execution):

- dist:clean: Clean the output directory, remove all existing files
- dist:fontawesome: Compile the Font Awesome Fonts and CSS files (2 sub-steps)
- dist:css-ltr: Compile CSS starting with `warehouse.scss` (left-to-right)
- dist:css-rtl: Compile CSS starting with `warehouse.scss` (right-to-left)
- dist:noscript: Compile CSS starting with `noscript.scss`
- dist:js: Compile JavaScript starting with `warehouse/static/js/warehouse/index.js`
  - this is the main JavaScript entry point, and currently uses WebPack plugin
- dist:admin:fonts: Compile Font Awesome Fonts for the admin interface
- dist:admin:css: Compile admin CSS. Has specific ordering of files to compile. Results in `all.css`
- dist:admin:js: Compile admin JavaScript to a bundle
- dist:vendor: Compile vendor JavaScript to a bundle named `vendor.js`
- dist:images: Compress images into the output directory
- dist:manifest: Generate a manifest file for the static assets, including cachebusting hashes
- dist:admin:manifest: Generate a manifest file for the admin static assets, including cachebusting hashes
- dist:compress:*: Multiple tasks, compress the static assets with gzip and brotli

I used this as a guideline to make sure I covered all cases.

## Proposed Implementation

The proposed implementation is to use Webpack to compile all static assets, and to use Webpack's plugin ecosystem to replace the Gulp plugins we currently use.

Instead of trying to replace parts of the `Gulpfile` piecemeal, approach this as a complete rewrite, and remove `gulp` entirely.

### Webpack Configuration

The Webpack configuration will be located in a `webpack.config.js` file in the root of the project.

### Webpack Plugins

Webpack plugins will be used to replace the Gulp plugins we currently use.

See `package.json` for the list of plugins we will use in the `devDependencies`.

---

## Implementation Plan

- [x] Remove `gulp` and `gulp-*` and other related packages from `package.json`
- [x] Add `webpack` and `webpack-cli` packages to `package.json` `devDependencies`
- [x] ~~Add eslint config for webpack https://github.com/webpack-contrib/eslint-config-webpack~~ - removed due to conflicting rules
- [x] Add `webpack` configuration file `webpack.config.js`
- [x] Configure compression plugins
- [x] Configure source map generation
- [x] Add `webpack` SCSS loaders to `package.json` `devDependencies` and config
- [x] Add Image minimization https://webpack.js.org/plugins/image-minimizer-webpack-plugin/
- [x] Add RTL support for CSS
- [x] Update `bin` scripts to use `webpack` instead of `gulp`
- [x] Add admin bundle configuration
- [x] Set mode https://webpack.js.org/guides/production/#specify-the-mode
- [x] Update docs to reflect new build process

### Deferred
_Things to take care of at a later date, but wanted to call out explicitly that I did not attempt to take these on right now._

- Update/migrate vendored `zxcvbn` library. It's been [rewritten in TypeScript](https://zxcvbn-ts.github.io/zxcvbn/guide/migration/). We should migrate to the new version and import the JS in our password controllers, and leverage webpack.
- Remove `babel` from the build process. We probably don't need it anymore, since we don't support IE11.
- Replace `imagemin` with `sharp` https://webpack.js.org/plugins/image-minimizer-webpack-plugin/#optimize-with-sharp
- Replace vendored bourbon & neat with npm CSS modules.
- Update of AdminLTE site to v3.2 and related assets
- Remove unused code from admin site, like `ionicons`
- Enable Dev-time hot module replacement

I wanted to complete this effort before taking any more. 

### Package stats

Before: 2023 packages. 55 vulnerabilities (19 moderate, 31 high, 5 critical)
After: 1622 packages. 28 vulnerabilities (14 moderate, 13 high, 1 critical)

--- 

# Testing

I spent a fair amount of time playing with the UI in a variety of ways to confirm behaviors, and while we don't have a full UI test yet, I confirmed that the JS behaviors continued to work, as well as the recompilation-on-change - which now takes a couple of seconds.

I could use help by folks pulling this down and running their own build/tests locally and confirming things work on par with production. The admin site is an area that I'm unfamiliar with it's current production behaviors, so I'd appreciate extra focus on that.

---

Refs: #6141, #7749, #12715

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+const sharedCompressionPlugins = [
+  new CompressionPlugin({
+    filename: "[path][base].gz",
+    algorithm: "gzip",
+    compressionOptions: { level: 9, memLevel: 9 },
+    // Only compress files that will actually be smaller when compressed.
+    minRatio: 1,
+  }),
+  /* TODO: Add plugins to compress brotli for text/font files vs the generic
+           Use BROTLI_MODE_TEXT/BROTLI_MODE_FONT and add a `test` qualifier. */
```
Comment (miketheman) [inline webpack.config.js:41]: I tinkered with this a bit locally and didn't find material differences - largely due to Brotli being generally great, and our files are relatively small, so the impact of change is in tens-of-bytes. I opted to keep the configuration simpler as a result.

**Review by di — APPROVED** [review]
I ran this locally and couldn't find any issues. The rebuild is crazy fast!

**Comment by di** [comment]
I noticed that there are now a lot of deprecation warnings in the container output, like:

```
static_1         |   WARNING in ./warehouse/static/sass/warehouse.scss (./warehouse/static/sass/warehouse.scss.webpack[javascript/auto]!=!./node_modules/css-loader/dist/cjs.js!./node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[0].oneOf[0].use[2]!./node_modules/sass-loader/dist/cjs.js!./warehouse/static/sass/warehouse.scss)
static_1         |   Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
static_1         |   Deprecation $weight: Passing a number without unit % (50) is deprecated.
static_1         |
static_1         |   To preserve current behavior: $weight * 1%
static_1         |
static_1         |   More info: https://sass-lang.com/d/function-units
static_1         |
static_1         |   warehouse/static/sass/blocks/_search-form.scss 52:12  @import
static_1         |   warehouse/static/sass/warehouse.scss 116:9            root stylesheet
static_1         |    @ ./warehouse/static/sass/warehouse.scss
static_1         |
static_1         |   WARNING in ./warehouse/static/sass/warehouse.scss (./warehouse/static/sass/warehouse.scss.webpack[javascript/auto]!=!./node_modules/css-loader/dist/cjs.js!./node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[0].oneOf[0].use[2]!./node_modules/sass-loader/dist/cjs.js!./warehouse/static/sass/warehouse.scss)
static_1         |   Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
static_1         |   88 repetitive deprecation warnings omitted.
static_1         |   Run in verbose mode to see all warnings.
static_1         |    @ ./warehouse/static/sass/warehouse.scss
static_1         |
static_1         |   WARNING in ./warehouse/static/sass/warehouse.scss (./warehouse/static/sass/warehouse.scss.webpack[javascript/auto]!=!./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./warehouse/static/sass/warehouse.scss)
static_1         |   Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
static_1         |   Deprecation $weight: Passing a number without unit % (50) is deprecated.
static_1         |
static_1         |   To preserve current behavior: $weight * 1%
static_1         |
static_1         |   More info: https://sass-lang.com/d/function-units
static_1         |
static_1         |   warehouse/static/sass/blocks/_search-form.scss 52:12  @import
static_1         |   warehouse/static/sass/warehouse.scss 116:9            root stylesheet
static_1         |    @ ./warehouse/static/sass/warehouse.scss
static_1         |
static_1         |   WARNING in ./warehouse/static/sass/warehouse.scss (./warehouse/static/sass/warehouse.scss.webpack[javascript/auto]!=!./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./warehouse/static/sass/warehouse.scss)
static_1         |   Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
static_1         |   88 repetitive deprecation warnings omitted.
static_1         |   Run in verbose mode to see all warnings.
static_1         |    @ ./warehouse/static/sass/warehouse.scss
static_1         |
static_1         |   4 warnings have detailed information that is not shown.
static_1         |   Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.
static_1         |
static_1         |   warehouse (webpack 5.75.0) compiled with 4 warnings in 2902 ms
```

Can we resolve these, so as not to clog up output in development, either in this PR or in a follow-on PR?

**Comment by miketheman** [comment]
I've tried to [remove some noise in the config](https://github.com/pypi/warehouse/pull/12728/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166fR103-R108) generated by the (now deprecated) Bourbon/Neat stack - we should pursue replacements as a follow up.

> Can we resolve these, so as not to clog up output in development

Done in aa13a1d5f27b1e4e666ac7601b17b8172ff927eb - we'll still retain 2 warnings (one per warehouse-* entry point) until we resolve all ~88 issues.

**Comment by miketheman** [comment]
> ... until we resolve all ~88 issues.

@di I have another set of commits to resolve these and get us a clean bill of health - but would want to ship that independently from this - how do you feel about that?

**Review by di — APPROVED** [review]

**Pull request merged**